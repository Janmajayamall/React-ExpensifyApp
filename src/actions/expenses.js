import uuid from 'uuid';
import database from './../firebase/firebase.js';
import { get } from 'http';

//Actions for Expenses 
export const addExpense = (expense) => ({
    type:'ADD_EXPENSE',
    expense: expense
});

export const startAddExpense = (expenseData = {}) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
            } = expenseData;
   
   const expense = {description, note, amount, createdAt};
          
   return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
     dispatch(addExpense({
         id: ref.key,
         ...expense
     }));
   });
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id: id
});

export const startRemoveExpense = (id) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense(id));
        });
    }
}

export const editExpense = (id, update)=>({
    type: 'EDIT_EXPENSE',
    id:id,
    update: update
});

export const startEditExpense = (id, update) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(update).then(()=>{
            dispatch(editExpense(id, update));
        });
    }
};



export const setExpenses = (expenses) =>({
    type: 'SET_EXPENSES',
    expenses: expenses
})

export const startSetExpenses = () => {
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
       return database.ref(`users/${uid}/expenses`).once("value").then((snapshot)=>{
            const expenses = [];

            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expenses));
        });
    };
};