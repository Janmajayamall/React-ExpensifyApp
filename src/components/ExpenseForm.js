import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';


export default class ExpenseForm extends React.Component{
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note:props.expense ? props.expense.note : '',
            amount:props.expense ? (props.expense.amount/100).toString() : '',
            createdAt:props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        };

    }
    
  

    onDescriptionChange = (e) =>{
        const description = e.target.value;
        this.setState(()=>{
            return {
                description: description
            }
        })
    }

    onNoteChange = (e) =>{
        const note = e.target.value;
        this.setState(()=>{
            return{
                note: note
            }
        })
    }

    onAmountChange = (e) =>{
        const amount = e.target.value;

    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>{
                return{
                    amount: amount
                }
            })
        }
       
    }

    onDateChange = (createdAt) =>{
        if(createdAt){
            this.setState(()=>({createdAt:createdAt}));

        }
    }

    onFocusChange = ({focused}) =>{
        this.setState(()=>({calenderFocused:focused}));
    }

    onSubmit = (e) =>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=>{
                return{
                    error: 'Description and Amount of the expense are required fields'
                }
            });
        }else{
            this.setState(()=>{
                return{
                    error:''
                }
            });
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount)*100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render(){
        return(
             <form className='form' onSubmit = {this.onSubmit}>
                    {this.state.error && <p className='form__error'>{this.state.error}</p>}
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value = {this.state.description}
                        className = "text-input"
                        onChange = {this.onDescriptionChange}

                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value = {this.state.amount}
                        className = "text-input"
                        onChange = {this.onAmountChange}

                    />
                    <textarea
                        className='text-area'
                        placeholder="If you want to add any notes on the expense, add them here!"
                        value = {this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                    <SingleDatePicker
                        date= {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = {this.state.calenderFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths= {1}
                        isOutsideRange= {()=>(false)}
                    />
                    <div>
                    <button className='button'>
                        Add Expense
                    </button>
                    </div>
                </form>
        );
    }
}