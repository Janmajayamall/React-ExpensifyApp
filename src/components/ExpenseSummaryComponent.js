import React from 'react';
import getTotalExpenses from './../selectors/expensesTotal.js';
import numeral from 'numeral'


export default class ExpenseSummaryComponent extends React.Component{
    render(){
        const expenseSING = this.props.expenses.length === 1 ? 'expense' : 'expenses';
        return(
            <div>
                <p>Viewing {this.props.expenses.length} {expenseSING} totalling {numeral(getTotalExpenses(this.props.expenses)/100).format('$0,0.00')}</p> 
            </div>
        );
    }
}

