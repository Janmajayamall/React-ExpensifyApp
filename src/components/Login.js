import React from 'react';
import {connect} from 'react-redux';
import {startLogin} from './../actions/auth.js'

export const LoginPage = (props) =>{
    return(
        <div className = 'box-layout'>
            <div className = 'box-layout__box'>
                <h1 className='box-layout__title'>SpendAdd</h1>
                <p>Don't worry for expenses from now onwards</p>
                <button className = 'button' onClick={props.startLogin}>Google Login</button>
            </div>
        </div>
    );
}

const mapDispatchToProps = (dispatch) =>{
    return {
        startLogin: () =>{dispatch(startLogin())}
    }
}

export default connect(undefined, mapDispatchToProps)(LoginPage);
