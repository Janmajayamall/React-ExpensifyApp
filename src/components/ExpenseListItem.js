import React from 'react';
import { connect } from 'react-redux';
import {removeExpense} from './../actions/expenses.js';
import {BrowserRouter, Route, Switch, Link, NavLink} from 'react-router-dom'


const ExpenseListItem = ( {dispatch ,id ,description, amount, createdAt}) =>{
    return(
        <div>
            <Link to={`/edit/${id}`} >{description}</Link>
            <p>Amount: {amount/100} and createdAt: {createdAt}</p>

        </div>
    );
}   

export default ExpenseListItem;