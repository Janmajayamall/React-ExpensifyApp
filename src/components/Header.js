import React from 'react';
import {BrowserRouter, Route, Switch, Link, NavLink} from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from './../actions/auth';

   

export const Header = (props)=>(
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName = "is-active" exact={true}> Go Home </NavLink>
        <NavLink to="/create" activeClassName = "is-active"> Add Expense</NavLink>
        <NavLink to="/help" activeClassName = "is-active"> Help </NavLink>
        <button onClick={props.startLogout}>Logout</button>
    </header>
);

const mapDispatchToProps = (dispatch) =>{
    return {
        startLogout: () =>{dispatch(startLogout())}
    }
}

export default connect(undefined, mapDispatchToProps)(Header);