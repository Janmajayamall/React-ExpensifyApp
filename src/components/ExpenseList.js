import React from 'react';
import ExpenseListItem from './ExpenseListItem';



export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? <p>No expense to show</p>
                : props.expenses.map((expense)=>{
            return(
                <ExpenseListItem key = {expense.id} {...expense}/>
                    );       
            })
        }
    </div>
);



export default ExpenseList;