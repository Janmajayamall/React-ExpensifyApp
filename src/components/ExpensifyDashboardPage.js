import React from 'react';
import {BrowserRouter, Route, Switch, Link, NavLink} from 'react-router-dom';
import {ExpenseList} from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters.js';
import ExpenseSummaryComponent from './ExpenseSummaryComponent.js';
import { connect } from 'react-redux';
import selectExpenses from './../selectors/expenses.js';


export const ExpensifyDashboardPage = (props) =>{

        return (
            <div>
                <ExpenseSummaryComponent expenses = {props.expenses} />
                <ExpenseListFilters/>
                <ExpenseList expenses = {props.expenses}/>
            </div>
        );}


const mapStateToProps = (state, props)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps)(ExpensifyDashboardPage);