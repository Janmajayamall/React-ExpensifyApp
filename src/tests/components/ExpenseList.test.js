import React from 'react';
import {ExpenseList} from './../../components/ExpenseList';
import {shallow} from 'enzyme';
import expenses from './../fixtures/expenses.js'

test('should render ExpenseList with expenses', ()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should rednder ExpenseList with empty message', ()=>{
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});