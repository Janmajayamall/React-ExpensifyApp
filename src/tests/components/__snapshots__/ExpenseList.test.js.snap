// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should rednder ExpenseList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No expense to show
      </span>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={2}
      createdAt={-345600000}
      description="gum"
      id={1}
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={0}
      createdAt={0}
      description="cell"
      id={2}
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={1}
      createdAt={345600000}
      description="bell"
      id={3}
      key="3"
      note=""
    />
  </div>
</div>
`;
